File: backend-python/app/utils/excel_parser.py
import re
from datetime import datetime
from pathlib import Path

from openpyxl import load_workbook

# Hebrew column name to field mapping
COLUMN_MAPPING = {
    "תאריך": "date",
    "תיאור": "description",
    "אמצעי תשלום": "payment_method",
    "קטגוריה": "category",
    "פירוט": "details",
    "אסמכתא": "reference",
    "חובה": "expense",
    "זכות": "income",
    "יתרה": "balance",
}

# Hebrew month names for parsing dates
HEBREW_MONTHS = {
    "ינו": 1,
    "ינואר": 1,
    "פבר": 2,
    "פברואר": 2,
    "מרץ": 3,
    "מרס": 3,
    "אפר": 4,
    "אפריל": 4,
    "מאי": 5,
    "יונ": 6,
    "יוני": 6,
    "יול": 7,
    "יולי": 7,
    "אוג": 8,
    "אוגוסט": 8,
    "ספט": 9,
    "ספטמבר": 9,
    "אוק": 10,
    "אוקטובר": 10,
    "נוב": 11,
    "נובמבר": 11,
    "דצמ": 12,
    "דצמבר": 12,
}

# English month abbreviations
ENGLISH_MONTHS = {
    "jan": 1,
    "january": 1,
    "feb": 2,
    "february": 2,
    "mar": 3,
    "march": 3,
    "apr": 4,
    "april": 4,
    "may": 5,
    "jun": 6,
    "june": 6,
    "jul": 7,
    "july": 7,
    "aug": 8,
    "august": 8,
    "sep": 9,
    "sept": 9,
    "september": 9,
    "oct": 10,
    "october": 10,
    "nov": 11,
    "november": 11,
    "dec": 12,
    "december": 12,
}


def parse_date(date_value) -> tuple[int | None, str]:
    """
    Parse date from various formats.
    Returns (unix_timestamp, raw_date_string).
    """
    if date_value is None:
        return None, ""

    raw_string = str(date_value).strip()

    # If it's already a datetime object
    if isinstance(date_value, datetime):
        return int(date_value.timestamp()), raw_string

    # Try to parse string formats
    date_str = raw_string.lower()

    # Try DD/MM/YYYY or DD-MM-YYYY
    match = re.match(r"(\d{1,2})[/\-](\d{1,2})[/\-](\d{2,4})", date_str)
    if match:
        day, month, year = int(match.group(1)), int(match.group(2)), int(match.group(3))
        if year < 100:
            year += 2000
        try:
            dt = datetime(year, month, day)
            return int(dt.timestamp()), raw_string
        except ValueError:
            pass

    # Try "DD Month YYYY" format (Hebrew or English)
    match = re.match(r"(\d{1,2})\s+([א-ת\w]+)\s+(\d{2,4})", date_str)
    if match:
        day = int(match.group(1))
        month_str = match.group(2).lower()
        year = int(match.group(3))
        if year < 100:
            year += 2000

        month = HEBREW_MONTHS.get(month_str) or ENGLISH_MONTHS.get(month_str)
        if month:
            try:
                dt = datetime(year, month, day)
                return int(dt.timestamp()), raw_string
            except ValueError:
                pass

    return None, raw_string


def parse_amount(value) -> float:
    """Parse amount from various formats (handles commas, spaces, etc)."""
    if value is None:
        return 0.0

    if isinstance(value, int | float):
        return float(value)

    # Remove quotes, commas, spaces
    cleaned = str(value).replace('"', "").replace(",", "").replace(" ", "").strip()

    if not cleaned or cleaned == "-":
        return 0.0

    try:
        return float(cleaned)
    except ValueError:
        return 0.0


def parse_excel_file(file_path: str | Path) -> dict[str, list[dict]]:
    """
    Parse an Excel file with multiple sheets.
    Each sheet represents an account.
    Returns dict with sheet name as key and list of transaction dicts as value.
    """
    wb = load_workbook(filename=file_path, data_only=True)
    result = {}

    for sheet_name in wb.sheetnames:
        sheet = wb[sheet_name]
        transactions = parse_sheet(sheet)
        if transactions:
            result[sheet_name] = transactions

    return result


def parse_sheet(sheet) -> list[dict]:
    """Parse a single Excel sheet into transaction dictionaries."""
    rows = list(sheet.iter_rows(values_only=True))
    if not rows:
        return []

    # Find header row by looking for Hebrew column names
    header_row_idx = None
    column_indices = {}

    for i, row in enumerate(rows):
        if row is None:
            continue

        for col_idx, cell in enumerate(row):
            if cell is None:
                continue
            cell_str = str(cell).strip()
            if cell_str in COLUMN_MAPPING:
                column_indices[col_idx] = COLUMN_MAPPING[cell_str]

        if column_indices:
            header_row_idx = i
            break

    if header_row_idx is None or not column_indices:
        return []

    # Parse data rows
    transactions = []
    for row in rows[header_row_idx + 1 :]:
        if row is None or all(cell is None for cell in row):
            continue

        txn = {}
        for col_idx, field_name in column_indices.items():
            if col_idx < len(row):
                value = row[col_idx]

                if field_name == "date":
                    timestamp, raw_string = parse_date(value)
                    txn["date"] = timestamp
                    txn["raw_date_string"] = raw_string
                elif field_name in ("expense", "income", "balance"):
                    txn[field_name] = parse_amount(value)
                else:
                    txn[field_name] = str(value).strip() if value else None

        # Skip rows without valid date or description
        if txn.get("date") is None or not txn.get("description"):
            continue

        # Set defaults for missing fields
        txn.setdefault("payment_method", None)
        txn.setdefault("category", "לא מסווג")  # "Uncategorized" in Hebrew
        txn.setdefault("details", None)
        txn.setdefault("reference", None)
        txn.setdefault("expense", 0.0)
        txn.setdefault("income", 0.0)
        txn.setdefault("balance", 0.0)

        transactions.append(txn)

    return transactions

File: backend/src/utils/excelParser.ts
import * as XLSX from "xlsx";

// Hebrew column name to field mapping
export const COLUMN_MAPPING: Record<string, string> = {
  תאריך: "date",
  תיאור: "description",
  "אמצעי תשלום": "payment_method",
  קטגוריה: "category",
  פירוט: "details",
  אסמכתא: "reference",
  חובה: "expense",
  זכות: "income",
  יתרה: "balance",
};

// Hebrew month names for parsing dates
const HEBREW_MONTHS: Record<string, number> = {
  ינו: 1,
  ינואר: 1,
  פבר: 2,
  פברואר: 2,
  מרץ: 3,
  מרס: 3,
  אפר: 4,
  אפריל: 4,
  מאי: 5,
  יונ: 6,
  יוני: 6,
  יול: 7,
  יולי: 7,
  אוג: 8,
  אוגוסט: 8,
  ספט: 9,
  ספטמבר: 9,
  אוק: 10,
  אוקטובר: 10,
  נוב: 11,
  נובמבר: 11,
  דצמ: 12,
  דצמבר: 12,
};

// English month abbreviations
const ENGLISH_MONTHS: Record<string, number> = {
  jan: 1,
  january: 1,
  feb: 2,
  february: 2,
  mar: 3,
  march: 3,
  apr: 4,
  april: 4,
  may: 5,
  jun: 6,
  june: 6,
  jul: 7,
  july: 7,
  aug: 8,
  august: 8,
  sep: 9,
  sept: 9,
  september: 9,
  oct: 10,
  october: 10,
  nov: 11,
  november: 11,
  dec: 12,
  december: 12,
};

export function parseDate(
  dateValue: unknown
): { timestamp: number | null; rawString: string } {
  if (dateValue == null) {
    return { timestamp: null, rawString: "" };
  }

  const rawString = String(dateValue).trim();

  // If it's a Date object
  if (dateValue instanceof Date) {
    return {
      timestamp: Math.floor(dateValue.getTime() / 1000),
      rawString,
    };
  }

  // If it's a number (Excel serial date)
  if (typeof dateValue === "number") {
    // Excel serial date conversion
    // Excel dates are days since 1899-12-30 (with a bug for 1900 leap year)
    const excelEpoch = new Date(1899, 11, 30);
    const msPerDay = 24 * 60 * 60 * 1000;
    const jsDate = new Date(excelEpoch.getTime() + dateValue * msPerDay);
    if (!isNaN(jsDate.getTime())) {
      return {
        timestamp: Math.floor(jsDate.getTime() / 1000),
        rawString,
      };
    }
  }

  const dateStr = rawString.toLowerCase();

  // Try DD/MM/YYYY or DD-MM-YYYY
  const slashMatch = dateStr.match(/(\d{1,2})[/\-](\d{1,2})[/\-](\d{2,4})/);
  if (slashMatch) {
    let [, dayStr, monthStr, yearStr] = slashMatch;
    let day = parseInt(dayStr, 10);
    let month = parseInt(monthStr, 10);
    let year = parseInt(yearStr, 10);

    if (year < 100) {
      year += 2000;
    }

    try {
      const date = new Date(year, month - 1, day);
      if (!isNaN(date.getTime())) {
        return {
          timestamp: Math.floor(date.getTime() / 1000),
          rawString,
        };
      }
    } catch {
      // Continue to next format
    }
  }

  // Try "DD Month YYYY" format (Hebrew or English)
  const textMatch = dateStr.match(/(\d{1,2})\s+([א-ת\w]+)\s+(\d{2,4})/);
  if (textMatch) {
    const day = parseInt(textMatch[1], 10);
    const monthStr = textMatch[2].toLowerCase();
    let year = parseInt(textMatch[3], 10);

    if (year < 100) {
      year += 2000;
    }

    const month = HEBREW_MONTHS[monthStr] ?? ENGLISH_MONTHS[monthStr];
    if (month) {
      try {
        const date = new Date(year, month - 1, day);
        if (!isNaN(date.getTime())) {
          return {
            timestamp: Math.floor(date.getTime() / 1000),
            rawString,
          };
        }
      } catch {
        // Return null timestamp
      }
    }
  }

  return { timestamp: null, rawString };
}

export function parseAmount(value: unknown): number {
  if (value == null) {
    return 0;
  }

  if (typeof value === "number") {
    return value;
  }

  // Remove quotes, commas, spaces
  const cleaned = String(value)
    .replace(/"/g, "")
    .replace(/,/g, "")
    .replace(/\s/g, "")
    .trim();

  if (!cleaned || cleaned === "-") {
    return 0;
  }

  const num = parseFloat(cleaned);
  return isNaN(num) ? 0 : num;
}

export interface ParsedTransaction {
  date: number | null;
  description: string;
  payment_method: string | null;
  category: string;
  details: string | null;
  reference: string | null;
  expense: number;
  income: number;
  balance: number;
  raw_date_string: string;
}

export function parseExcelFile(
  buffer: Buffer
): Record<string, ParsedTransaction[]> {
  const workbook = XLSX.read(buffer, { type: "buffer" });
  const result: Record<string, ParsedTransaction[]> = {};

  for (const sheetName of workbook.SheetNames) {
    const sheet = workbook.Sheets[sheetName];
    const transactions = parseSheet(sheet);
    if (transactions.length > 0) {
      result[sheetName] = transactions;
    }
  }

  return result;
}

export function parseSheet(
  sheet: XLSX.WorkSheet
): ParsedTransaction[] {
  const rows = XLSX.utils.sheet_to_json(sheet, {
    header: 1,
    raw: true,
  }) as unknown[][];

  if (!rows || rows.length === 0) {
    return [];
  }

  // Find header row by looking for Hebrew column names
  let headerRowIdx: number | null = null;
  const columnIndices: Map<number, string> = new Map();

  for (let i = 0; i < rows.length; i++) {
    const row = rows[i];
    if (!row) continue;

    for (let colIdx = 0; colIdx < row.length; colIdx++) {
      const cell = row[colIdx];
      if (cell == null) continue;

      const cellStr = String(cell).trim();
      if (cellStr in COLUMN_MAPPING) {
        columnIndices.set(colIdx, COLUMN_MAPPING[cellStr]);
      }
    }

    if (columnIndices.size > 0) {
      headerRowIdx = i;
      break;
    }
  }

  if (headerRowIdx === null || columnIndices.size === 0) {
    return [];
  }

  // Parse data rows
  const transactions: ParsedTransaction[] = [];

  for (let i = headerRowIdx + 1; i < rows.length; i++) {
    const row = rows[i];
    if (!row || row.every((cell) => cell == null)) {
      continue;
    }

    const txn: Partial<ParsedTransaction> = {};

    for (const [colIdx, fieldName] of columnIndices) {
      const value = colIdx < row.length ? row[colIdx] : null;

      if (fieldName === "date") {
        const { timestamp, rawString } = parseDate(value);
        txn.date = timestamp;
        txn.raw_date_string = rawString;
      } else if (["expense", "income", "balance"].includes(fieldName)) {
        (txn as Record<string, number>)[fieldName] = parseAmount(value);
      } else {
        (txn as Record<string, string | null>)[fieldName] =
          value != null ? String(value).trim() : null;
      }
    }

    // Skip rows without valid date or description
    if (txn.date == null || !txn.description) {
      continue;
    }

    // Set defaults for missing fields
    transactions.push({
      date: txn.date,
      description: txn.description!,
      payment_method: txn.payment_method ?? null,
      category: txn.category ?? "לא מסווג",
      details: txn.details ?? null,
      reference: txn.reference ?? null,
      expense: txn.expense ?? 0,
      income: txn.income ?? 0,
      balance: txn.balance ?? 0,
      raw_date_string: txn.raw_date_string ?? "",
    });
  }

  return transactions;
}

export function getSheetPreview(
  buffer: Buffer,
  sheetName: string
): {
  headers: string[];
  detectedMapping: Record<string, string>;
  rowCount: number;
  sampleRows: Record<string, string>[];
} | null {
  const workbook = XLSX.read(buffer, { type: "buffer" });

  if (!workbook.SheetNames.includes(sheetName)) {
    return null;
  }

  const sheet = workbook.Sheets[sheetName];
  const rows = XLSX.utils.sheet_to_json(sheet, {
    header: 1,
    raw: true,
  }) as unknown[][];

  if (!rows || rows.length === 0) {
    return null;
  }

  // Find headers and mapping
  const headers: string[] = [];
  const detectedMapping: Record<string, string> = {};
  let headerRowIdx: number | null = null;
  const headerColIndices: number[] = [];

  for (let i = 0; i < rows.length; i++) {
    const row = rows[i];
    if (!row) continue;

    for (let colIdx = 0; colIdx < row.length; colIdx++) {
      const cell = row[colIdx];
      if (cell == null) continue;

      const cellStr = String(cell).trim();
      if (cellStr in COLUMN_MAPPING) {
        headers.push(cellStr);
        headerColIndices.push(colIdx);
        detectedMapping[cellStr] = COLUMN_MAPPING[cellStr];
      }
    }

    if (headers.length > 0) {
      headerRowIdx = i;
      break;
    }
  }

  if (headerRowIdx === null) {
    return null;
  }

  // Count data rows
  const dataRows = rows.slice(headerRowIdx + 1);
  const rowCount = dataRows.filter(
    (row) => row && !row.every((cell) => cell == null)
  ).length;

  // Get sample rows (first 3)
  const sampleRows: Record<string, string>[] = [];
  for (const row of dataRows.slice(0, 3)) {
    if (!row || row.every((cell) => cell == null)) continue;

    const sample: Record<string, string> = {};
    for (let i = 0; i < headers.length; i++) {
      const colIdx = headerColIndices[i];
      const val = colIdx < row.length ? row[colIdx] : null;
      sample[headers[i]] = val != null ? String(val) : "";
    }
    sampleRows.push(sample);
  }

  return {
    headers,
    detectedMapping,
    rowCount,
    sampleRows,
  };
}

export function getWorkbookSheetNames(buffer: Buffer): string[] {
  const workbook = XLSX.read(buffer, { type: "buffer" });
  return workbook.SheetNames;
}
